satellite/gracefulexit/endpoint_test.go

```diff
@@ -367,7 +366,120 @@ func TestFailureUplinkSignature(t *testing.T) {
+
+					header, err := pieceReader.GetPieceHeader()
+					require.NoError(t, err)
+
+					orderLimit := header.OrderLimit
+					originalPieceHash := &pb.PieceHash{
+						PieceId:   orderLimit.PieceId,
+						Hash:      header.GetHash(),
+						PieceSize: pieceReader.Size(),
+						Timestamp: header.GetCreationTime(),
+						Signature: header.GetSignature(),
+					}
+
+					newPieceHash := &pb.PieceHash{
+						PieceId:   m.TransferPiece.AddressedOrderLimit.Limit.PieceId,
+						Hash:      originalPieceHash.Hash,
+						PieceSize: originalPieceHash.PieceSize,
+						Timestamp: time.Now(),
+					}
+
+					receivingNode := storageNodes[m.TransferPiece.AddressedOrderLimit.Limit.StorageNodeId]
+					require.NotNil(t, receivingNode)
+
+					// get the receiving node piece count before processing
+					recNodeID = receivingNode.ID()
+					recNodePieceCount, err = getNodePieceCount(ctx, satellite, recNodeID, numObjects)
+					require.NoError(t, err)
+
+					signer := signing.SignerFromFullIdentity(receivingNode.Identity)
+
+					signedNewPieceHash, err := signing.SignPieceHash(ctx, signer, newPieceHash)
+					require.NoError(t, err)
+
+					success := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Succeeded{
+							Succeeded: &pb.TransferSucceeded{
+								OriginalPieceId:      m.TransferPiece.OriginalPieceId,
+								OriginalPieceHash:    originalPieceHash,
+								OriginalOrderLimit:   &orderLimit,
+								ReplacementPieceHash: signedNewPieceHash,
+							},
+						},
+					}
+					err = processClient.Send(success)
+					require.NoError(t, err)
+				} else {
+					failedCount++
+					failed := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Failed{
+							Failed: &pb.TransferFailed{
+								OriginalPieceId: m.TransferPiece.OriginalPieceId,
+								Error:           pb.TransferFailed_UNKNOWN,
+							},
+						},
+					}
+					err = processClient.Send(failed)
+					require.NoError(t, err)
+				}
+			case *pb.SatelliteMessage_ExitCompleted:
+				// TODO test completed signature stuff
+				break
+			default:
+
+			}
+		}
+
+		// check that the exit has completed and we have the correct transferred/failed values
+		progress, err := satellite.DB.GracefulExit().GetProgress(ctx, exitingNode.ID())
+		require.NoError(t, err)
+
+		require.EqualValues(t, numPieces, progress.PiecesTransferred)
+		// even though we failed 1, it eventually succeeded, so the count should be 0
+		require.EqualValues(t, 0, progress.PiecesFailed)
+
+		// get the exiting node piece count after processing
+		exitNodePieceCountLatest, err := getNodePieceCount(ctx, satellite, exitingNode.ID(), numObjects*2)
+		require.NoError(t, err)
+		require.Equal(t, 0, exitNodePieceCountLatest)
+
+		// get the receiving node piece count after processing
+		recNodePieceCountLatest, err := getNodePieceCount(ctx, satellite, recNodeID, numObjects*2)
+		require.NoError(t, err)
+		require.Equal(t, recNodePieceCount+1, recNodePieceCountLatest)
```

Comment:
We should make sure to get the remote segment from metainfo, make sure it contains `recNodeID` and does not contain `exitingNodeID`

Reply:
Updated